### ISSUE TITLE
autoload missing for which-key-setup-side-window-right-bottom

### ISSUE THREAD

**Issue by mbertheau** [justbur/emacs-which-key#60]
I believe there's an autoload missing for `which-key-setup-side-window-right-bottom`

https://github.com/justbur/emacs-which-key/blob/156bd61f46eb054ccbd863539d05c7e5d892b321/which-key.el#L418

Also maybe an option to have which-key appear on the left side could be added.

**Comment by justbur** [comment]
Yep thanks!

It's an option already. Take a look at which-key-side-window-location.

**Comment by mbertheau** [comment]
@justbur Thanks for fixing so quickly!

option for left side: Ah, yes, of course. I was thinking about spacemacs.

**Comment by justbur** [comment]
@mbertheau I'm uncomfortable with making it one of the "default" options, because it's disorienting to me to have the buffer popup on the left and push all of the text to the right. Maybe you disagree?

**Comment by mbertheau** [comment]
@justbur just usually when I'm entering a SPC command, my eyes are in the bottom left. helm buffers open there, the minibuffer is there with `M-:` and so on. With which-key on the right, and the text top-aligned, I have to look at the corner the farthest from where the action is.

I'm running with a left-then-bottom configuration now, but not long enough to find if I like it. I just found that the list of keys is hard to parse for me at the bottom. Usually there would be 2-3 lines of keys with maybe 7 or 8 keys on one line, so there was little advantage when scanning the list from using a table to lay out the keys.

**Comment by justbur** [comment]
I'm open to suggestions if you want to propose some different formatting
for the buffer. I think I have the layout functions down so changing the
formatting of the keys should not be an issue. I could add options for
padding or something like that.

**Comment by justbur** [comment]
I think my current suggestion for the problem of not being able to find keys is to use the new option `which-key-prevent-C-h-from-cycling` (along with `which-key-use-C-h-for-paging`). What this does is to allow you to page through the keys using `C-h` and then drop into `helm-descbinds` after the last page. So you can use helm to narrow down choices if you don't find it scanning the keys. 

Other possibilities (within the current version) are to play with different sorting functions (there's one to sort on the command description) and the max dimensions. I think it might be hard to fully "solve" this problem, as there is always the possibility of having prefixes with many, many bindings. 

guide-key's solution was to just not show them when there were too many of them :-)

**Comment by mbertheau** [comment]
The problem for me is not that it's many keys, but that they are laid out in a way that makes it hard for me to scan them. A window on the left or right is high enough that the keys appear in one column one under the other. In a window on the bottom it's often 7 to 8 columns with only one or two rows each, which is much harder to scan quickly.

I'd like to try either a higher bottom window so that I have no more than 2 or 3 columns even with many keys, or a left window with the keys aligned to the bottom of the window.

**Comment by mbertheau** [comment]
That also opens the possibility to show longer command names, which sometimes disambiguates things, for example in `SPC k`

**Comment by justbur** [comment]
For `SPC k`, let's do 

```elisp
(add-to-list 'which-key-description-replacement-alist '("\\`evil-lisp-state-" . ""))
```

I'm happy to provide an option to either limit the number of columns shown or add padding between columns, which would have the same effect. Is that something you want?

**Comment by justbur** [comment]
Is it the case that you know what you are looking for and can't find it quickly? Or are you simply "discovering" the bindings?

**Comment by justbur** [comment]
Try this and see if it makes it easier to find bindings
```elisp
(setq which-key-sort-order 'which-key-description-order)
```

**Comment by mbertheau** [comment]
It's both. Discovering definitely very often, but also looking for the binding when I know what prefix the command is in. When I don't remember or know the prefix I use `SPC h d b`.

What do you think about a minimum height for a bottom window? I.e. start a new column only when you have at least n keys in the current column. I'd try n = 15 maybe at the beginning. Extra padding between the columns spreads them out over wide screens and seems like an indirect measure to address the problem I'm having. Limiting the number of columns doesn't help much with few bindings and maybe forces the key list to be very tall for long key lists.

**Comment by mbertheau** [comment]
Hmm, a key to toggle between filtering the list of keys and descriptions and invoking the command would maybe also help with finding quickly.

**Comment by justbur** [comment]
I can add a minimum height. I have no problem with that. The toggle is interesting, but it's more complicated. I would think it would make the most sense to pick your preferred order and then drop into helm-descbinds if you can't find what you're looking for.

**Comment by mbertheau** [comment]
Yeah, I agree, duplicating helm-descbinds functionality doesn't make much sense.

Just wanted to say: which-key is great! :)

**Comment by justbur** [comment]
Thanks!

Thinking about it a little more, I'm not sure how a min height would work. If you only had one binding, would you still show a big mostly empty window?

My algorithm for choosing the number of columns and keys is roughly to start with columns that have the max height and then progressively lower the max height until I get to the point where lowering it any more would push keys onto a new page or we reach 1 line. I think what you want to do is to stop this process earlier than it stops now. That's easier to accomplish than the min height thing, because the min height option seems to have more special cases to consider. 

I could make it an option to stop at 2 lines instead of 1 for example.

**Comment by justbur** [comment]
Or if you wanted to get fancy you could give me a function to evaluate when to stop.

**Comment by mbertheau** [comment]
Yes. I wasn't very good at describing what I meant, sorry :)

What I meant is: I'd start with n (15) rows per column, and if the keys fit on one page, use that. If they don't fit on one page, increase n until they do.

**Comment by justbur** [comment]
@mbertheau I just pushed `which-key-min-display-lines`, which should work like you just described.

**Comment by mbertheau** [comment]
In other words, I don't want to go to the smallest number of lines, but instead to the smallest number of columns, as long as these columns are smaller than a reasonable height of n (maybe around 15).

**Comment by mbertheau** [comment]
If I had one binding the window would be high enough but not higher to accomodate 1 line.

**Comment by justbur** [comment]
Try `which-key-min-display-lines` at 15

**Comment by mbertheau** [comment]
Let me figure out how to install a package from github :)

**Comment by justbur** [comment]
There's no rush if you want to wait for melpa. It should only take at most a couple of hours to show up

**Comment by mbertheau** [comment]
Your algorithm is equivalent of course, indeed :)

**Comment by justbur** [comment]
On the other hand if you just download the which-key.el file and load it (or eval the buffer) that's all you need to do.

**Comment by justbur** [comment]
That was quick. The version on melpa is correct
